airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbRecordConsumer.java

```diff
@@ -0,0 +1,164 @@
+  private final MongoDatabase mongoDatabase;
+  private final ConfiguredAirbyteCatalog catalog;
+  private final Consumer<AirbyteMessage> outputRecordCollector;
+  private final ObjectMapper objectMapper;
+
+  private AirbyteMessage lastStateMessage = null;
+
+  public MongodbRecordConsumer(Map<AirbyteStreamNameNamespacePair, MongodbWriteConfig> writeConfigs,
+                               MongoDatabase mongoDatabase,
+                               ConfiguredAirbyteCatalog catalog,
+                               Consumer<AirbyteMessage> outputRecordCollector) {
+    this.writeConfigs = writeConfigs;
+    this.mongoDatabase = mongoDatabase;
+    this.catalog = catalog;
+    this.outputRecordCollector = outputRecordCollector;
+    this.objectMapper = new ObjectMapper();
+  }
+
+  @Override
+  protected void startTracked() {
+    // todo (cgardens) - move contents of #write into this method.
+  }
+
+  @Override
+  protected void acceptTracked(AirbyteMessage message) {
+    if (message.getType() == AirbyteMessage.Type.STATE) {
+      lastStateMessage = message;
+    } else if (message.getType() == AirbyteMessage.Type.RECORD) {
+      final AirbyteRecordMessage recordMessage = message.getRecord();
+      AirbyteStreamNameNamespacePair pair = AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage);
+
+      if (!writeConfigs.containsKey(pair)) {
+        throw new IllegalArgumentException(
+            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+                Jsons.serialize(catalog), Jsons.serialize(recordMessage)));
+      }
+      try {
+        Map<String, Object> result = objectMapper.convertValue(recordMessage.getData(), new TypeReference<>() {});
+        var newDocument = new Document();
+        newDocument.put(AIRBYTE_DATA, new Document(result));
+        newDocument.put(AIRBYTE_EMITTED_AT, new LocalDateTime().toString());
+
+        var writeConfig = writeConfigs.get(pair);
+        var collection = writeConfig.getCollection();
+
+        var documents = writeConfig.getDocuments();
+        if (!documents.contains(newDocument.get(AIRBYTE_DATA, Document.class))) {
+          collection.insertOne(newDocument);
+        }
+      } catch (RuntimeException e) {
+        LOGGER.error("Got an error while writing message:" + e.getMessage());
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  @Override
+  protected void close(boolean hasFailed) {
+    try {
+      if (!hasFailed) {
+        LOGGER.info("Migration finished with no explicit errors. Copying data from tmp tables to permanent");
+        writeConfigs.values()
+            .forEach(
+                mongodbWriteConfig -> copyTable(mongoDatabase, mongodbWriteConfig.getCollectionName(),
+                    mongodbWriteConfig.getTmpCollectionName()));
+        outputRecordCollector.accept(lastStateMessage);
+      } else {
+        LOGGER.warn("Had errors while migrations");
```

Comment:
Are you sure that this sure be a warn level? Pretty often in connects we write a final part of the migration and create actual records. So it if fails here - in most of the connectors that would mean a show-stopper issue and customers will not see migrated data.  
Also please consider the Io and runtime, etc excpeptions handling while finalizing migration. If it fails here - we would also got a currpted migration with no explicit reason in logs. Thanks

Reply:
updated it